### PR TITLE
Document FilterableList Search

### PR DESCRIPTION
**Pull request by alexm118**
A quick one page primer on how FilterableLists work, the core classes involved and the logic pertaining to them.

---

## Additions

- Documentation

## Checklist

- [ ] PR has an informative and human-readable title
  - PR titles are used to generate the change log in [releases](../../releases); good ones make that easier to scan.
  - Consider prefixing, e.g., "Mega Menu: fix layout bug", or "Docs: Update Docker installation instructions".
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code follows the standards laid out in the [CFPB development guidelines](https://github.com/cfpb/development)
- [ ] Future todos are captured in comments and/or tickets
- [ ] Project documentation has been updated, potentially one or more of:
  - [This repo’s docs](https://cfpb.github.io/consumerfinance.gov/) (edit the files in the `/docs` folder) – for basic, close-to-the-code docs on working with this repo
  - CFGOV/platform wiki on GHE – for internal CFPB developer guidance
  - CFPB/hubcap wiki on GHE – for internal CFPB design and content guidance

### CODE REVIEW

**Review by willbarton — APPROVED** [review]
This looks good to me on a read-through!

File: docs/filterable-lists.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+### Documents
+
+There is currently only one type of document defined `FilterablePagesDocument` which is based off the `AbstractFilterPage` class. This document is responsible for housing data related to any of the filterable page types that extend AbstractFilterPage, including EnforcementActions, BlogPage, EventPage, and NewsRoomPage to name a few. In order to get fields that are specific to a page type, such as the status list for an Enforcement Action, you use the `prepare_field` function syntax. The use of `get_instances_from_related` is to enforce the auto updating of our index when changes occur to a specific page we have indexed rather than just the relation to `AbstractFilterPage` that is reflected in the database.
+
+### Search
+
+Search is the final piece of the puzzle, where we actually leverage a couple elasticsearch features to filter and match documents and return them in an ordered queryset. Before breaking down the search classes, its important to discuss the current implementation from an Elasticsearch perspective to understand how we're gathering results.
```
Comment (Scotchester) [inline docs/filterable-lists.md:52]:
```suggestion
Search is the final piece of the puzzle, where we actually leverage Elasticsearch to filter and match documents and return them in an ordered `QuerySet`. Before breaking down the search classes, it's important to discuss the current implementation from an Elasticsearch perspective to understand how we're gathering results.
```

File: docs/filterable-lists.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+The expanded search for filterable lists is using a [multi-match](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-multi-match-query.html) across the title, topic name, preview description, and content fields of all FilterablePagesDocument. We are leveraging a [phrase_prefix](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-multi-match-query.html#type-phrase) matching style with a currently configured slop of 2, to allow for some looser matching restrictions. We also provide a boost score for matching to the title and topic name fields indicated with `^10` within the code base. This boost score is to enable better ordering by relevance when desired. Search currently supports two different methods of ordering results, relevance and date published. Relevance is calculated by the Elasticsearch engine when returning results, and the date published is calculated based on page publication date. Enforcement Actions define their own ordering logic based on initial filing date.
```
Comment (Scotchester) [inline docs/filterable-lists.md:54]: Is "expanded search" a term of art? Not fully sure I understand what that means.

```suggestion
The expanded search for filterable lists is using a [multi-match query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-multi-match-query.html) across the title, topic name, preview description, and content fields of all `FilterablePagesDocument`s. We are leveraging a [phrase_prefix](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-multi-match-query.html#type-phrase) matching style with a currently configured slop of 2, to allow for some looser matching restrictions. We also provide a boost score for matching to the title and topic name fields, indicated by `^10` within the code base. This boost score is to enable better ordering by relevance when desired. Search currently supports two different methods of ordering results: relevance and date published. Relevance is calculated by the Elasticsearch engine when returning results, and the date published is calculated based on page publication date. Enforcement Actions define their own ordering logic based on initial filing date. 
```

Comment (Scotchester) [reply docs/filterable-lists.md:54]: Also on this paragraph:
1. I don't see where the preview description, and content fields are configured for `FilterablePagesDocument`.
2. I also don't see the `^10` indicator in the codebase. I see `boost` params of 10 on non-`FilterablePagesDocument` document models, but not on `FilterablePagesDocument` itself.

File: docs/filterable-lists.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+#### FilterablePagesDocumentSearch
+
+FilterablePagesDocumentSearch is the core search class that is used across the majority of our searching. It is invoked from FilterableListForm. This search class defines the common structure for our search function, as well as the base logic for filtering against all common fields and logic behind our multi-match and ordering steps. The core function called from outside the class is the `search` function which properly chains all of our filter/match/sorting logic and returns the resulting list as a django queryset.
```
Comment (Scotchester) [inline docs/filterable-lists.md:58]:
```suggestion
`FilterablePagesDocumentSearch` is the core search class that is used across the majority of our searching. It is invoked from `FilterableListForm`. This search class defines the common structure for our search function, as well as the base logic for filtering against all common fields and logic behind our multi-match and ordering steps. The core function called from outside the class is the `search` function, which properly chains all of our filter/match/sorting logic and returns the resulting list as a Django `QuerySet`.
```

File: docs/filterable-lists.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+#### EventFilterablePagesDocumentSearch
+
+EventFilterablePagesDocumentSearch is an extension of FilterablePagesDocumentSearch that defines behavior specific to our Event search. The class overwrites one method from its parent, the `filter_date` function to change the behavior to filter based on fields specific to events, the start and end date of an event.
```
Comment (Scotchester) [inline docs/filterable-lists.md:62]:
```suggestion
`EventFilterablePagesDocumentSearch` is an extension of `FilterablePagesDocumentSearch` that defines behavior specific to our future and past Events listings. The class overwrites one method from its parent, the `filter_date` function, to change the behavior to filter based on fields specific to events, the start and end date of an event.
```

File: docs/filterable-lists.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+#### EnforcementActionsFilterableListForm
+
+EnforcementActionsFilterableListForm is an extension of FilterablePagesDocumentSearch that exposes some additional filter logic through the `apply_specific_filters` function. We also see that `filter_date` and `order_results` have been overwritten to leverage the enforcement action specific fields, initial filing date.
```
Comment (Scotchester) [inline docs/filterable-lists.md:66]:
```suggestion
`EnforcementActionsFilterableListForm` is an extension of `FilterablePagesDocumentSearch` that exposes some additional filter logic through the `apply_specific_filters` function. We also see that `filter_date` and `order_results` have been overwritten to leverage an Enforcement Action-specific field, initial filing date.
```

**Review by Scotchester — COMMENTED** [review]
This is a great doc, thank you @alexm118! Most of this is minor copyediting, but there are a couple more substantial questions in there.

File: docs/filterable-lists.md [hunk 6]
```diff
@@ -1,66 +1,66 @@
...
 ### CategoryFilterableListMixin
 
-The [CategoryFilterableListMixin](https://github.com/cfpb/consumerfinance.gov/blob/main/cfgov/v1/models/filterable_list_mixins.py#L199) is an extension of the base FilterableListMixin that exposes some new functionality. It modifies how `get_filterable_queryset` operates and that it gets an initial list of pages but limits them to only ones that also fall within a set of initial categories which can be definited as the variable `filterable_categories` on a given page. We can see this in action with both `Newsroom` and `Recent Updates` filterable pages.
+The [CategoryFilterableListMixin](https://github.com/cfpb/consumerfinance.gov/blob/main/cfgov/v1/models/filterable_list_mixins.py#L199) is an extension of the base `FilterableListMixin` that exposes some new functionality. It modifies how `get_filterable_queryset` operates in that it gets an initial list of pages but limits them to only ones that are assigned a category within a set of initial categories, which is defined as the variable `filterable_categories` on a given page model. We can see this in action with both Newsroom (`NewsroomLandingPage`) and Recent Updates (`ActivityLogPage`) pages.
```
Comment (higs4281) [inline docs/filterable-lists.md:28]:
```suggestion
The [CategoryFilterableMixin](https://github.com/cfpb/consumerfinance.gov/blob/main/cfgov/v1/models/filterable_list_mixins.py#L199) is an extension of the base `FilterableListMixin` that exposes some new functionality. It modifies how `get_filterable_queryset` operates in that it gets an initial list of pages but limits them to only ones that are assigned a category within a set of initial categories, which is defined as the variable `filterable_categories` on a given page model. We can see this in action with both Newsroom (`NewsroomLandingPage`) and Recent Updates (`ActivityLogPage`) pages.
```

Comment (higs4281) [reply docs/filterable-lists.md]: Nit: name mismatch.

File: docs/filterable-lists.md [hunk 7]
```diff
@@ -1,66 +1,66 @@
...
+This is the base form that the vast majority of cf.gov uses for filterable lists. It defines the core fields that are visible on the form as well as functions to assist in setting initial data and sanitizing form input. The important information regarding `FilterableListForm` is that it defines the function `get_page_set`, which is responsible for invoking a search query. The logic regarding how to pass categories into the search object is due to the previously mentioned `CategoryFilterableListMixin`, which modifies the initial search parameters to enforce a category search if and only if the `filterable_categories` list is passed into the form when initialized.
```
Comment (higs4281) [inline docs/filterable-lists.md:36]:
```suggestion
This is the base form that the vast majority of cf.gov uses for filterable lists. It defines the core fields that are visible on the form as well as functions to assist in setting initial data and sanitizing form input. The important information regarding `FilterableListForm` is that it defines the function `get_page_set`, which is responsible for invoking a search query. The logic regarding how to pass categories into the search object is due to the previously mentioned `CategoryFilterableMixin`, which modifies the initial search parameters to enforce a category search if and only if the `filterable_categories` list is passed into the form when initialized.
```

Comment (higs4281) [reply docs/filterable-lists.md:36]: same nit as above.

File: docs/filterable-lists.md (hunk 6 above)
Comment (Scotchester) [reply docs/filterable-lists.md:28]: Good catch!

File: docs/filterable-lists.md [hunk 8]
```diff
@@ -1,66 +1,66 @@
...
+`EventFilterablePagesDocumentSearch` is an extension of `FilterablePagesDocumentSearch` that defines behavior specific to our future and past Events listings. The class overwrites one method from its parent, the `filter_date` function, to change the behavior to filter based on fields specific to events, the start and end date of an event.
 
 #### EnforcementActionsFilterableListForm
 
-EnforcementActionsFilterableListForm is an extension of FilterablePagesDocumentSearch that exposes some additional filter logic through the `apply_specific_filters` function. We also see that `filter_date` and `order_results` have been overwritten to leverage the enforcement action specific fields, initial filing date.
\ No newline at end of file
+`EnforcementActionsFilterableListForm` is an extension of `FilterablePagesDocumentSearch` that exposes some additional filter logic through the `apply_specific_filters` function. We also see that `filter_date` and `order_results` have been overwritten to leverage an Enforcement Action-specific field, initial filing date.
```
Comment (higs4281) [inline docs/filterable-lists.md:66]:
```suggestion
`EnforcementActionsFilterForm` is an extension of `FilterablePagesDocumentSearch` that exposes some additional filter logic through the `apply_specific_filters` function. We also see that `filter_date` and `order_results` have been overwritten to leverage an Enforcement Action-specific field, initial filing date.
```

Comment (higs4281) [reply docs/filterable-lists.md:66]: Name slightly off.

**Review by higs4281 — APPROVED** [review]
Three nits on mixin names.

**Review by higs4281 — CHANGES_REQUESTED** [review]
Just need mixin name fixes

File: docs/filterable-lists.md (hunk 2 above)
Comment (Scotchester) [reply docs/filterable-lists.md]: @alexm118 Sorry to bug you, but if you have a minute to give a quick response to these questions, I'd be much obliged!

**Review by cwdavies — APPROVED** [review]
I verified that the FilterableList Mixin names have been fixed

File: docs/filterable-lists.md (hunk 2 above)
Comment (alexm118) [reply docs/filterable-lists.md]: Sorry! Just saw these. 

1. Expanded search was the term we were using to discusses enhancing the search across multiple fields. It doesn't have any specific meaning within elasticsearch.

2. [Boosting within the query](https://github.com/cfpb/consumerfinance.gov/pull/6320/files#diff-e67e781c1ded8ed3bb0a2b0c49b7f1664beb0b8678a9b05b20c0cefc4c605e3eR156) was implemented on #6320 which looks like it was just merged a few days ago.

**Review by higs4281 — APPROVED** [review]

File: docs/filterable-lists.md (hunk 2 above)
Comment (Scotchester) [reply docs/filterable-lists.md]: Thanks so much for clarifying, Alex!

**Pull request merged**